# Dzień pierwszy DWthon - Hack Outside the Box 
### Cel na dziś: wczytanie i wstępna analiza danych 

Witaj na DWthon - *hack outside the box* 🤝  Jeśli czujesz głód wiedzy i czujesz, że chcesz się rozwijać szybciej i bardziej efektywnie niż do tej pory, to jesteś we właściwym miejscu 🧠 Moją rolą jest pokazać Ci, jak krok po kroku można wydobyć wartość z danych i uczenia maszynowego. 👀    

Pamiętaj, że DWthon trwa 5 dni i nie jest naszym celem zagłębianie się w szczegóły. Celem jest pokazanie Ci, jak postawić pierwsze kroki ku nowej dziedzinie albo nowemu sposobowi myślenia, jeśli zagadnienia techniczne nie są dla Ciebie już zagadką. Z mojego doświadczenia wynika, że czasem nawet osoby zaawansowane technicznie nie działają optymalnie i popadają w tak zwane "czarne dziury" (tak w DataWorkshop nazywamy zadania, które pochłaniają bardzo dużo czasu, energii i pieniędzy, a nie zwracają nic konkretnego ;)). W przypadku projektów Machine Learning czy Data Science (akurat w obu przypadkach jest pełno czarnych dziur) można bardzo łatwo popaść w nadmiar pracy i efektu nie będzie lub będzie na tyle nieznaczny, że po podsumowaniu godzin spędzonych nad projektem a dostarczoną wartością bilans będzie raczej powodował smutek. Tak nie musi być! Właśnie po to powstał DWthon, aby pokazać pewien sposób myślenia (!) i ścieżkę działania, która powoduje, że stosunkowo małym wysiłkiem można osiągnąć całkiem sporo. 


Często odwołuję się w swoich [webinarach](https://bit.ly/3eg9XXM), czy [podcastach](https://bit.ly/3bqocaA) **Biznes Myśli** do **zasady Pareto - 80/20**, która w dużym skrócie mówi o tym, że 80% czegoś (tu może wstawić prawie wszystko, co chcesz np. wartość, energia, straty itd) generowane jest przez 20% czegoś (np. wysiłku, klientów itd.) 


Zadań, których możesz zrobić jest mnóstwo. Szczególików, nad którymi możesz się skupiać (szczególnie w Machine Learning) jest mnóstwo. Czy warto zaczynać od szczególików i rzeźbienia detali miesiącami, jak można w pierwszej kolejności zerwać nisko wiszące owoce, a dopiero potem, jak to już daje nam konkretną wartość, skupić się na detalach, jeśli faktycznie 1% robi różnicę (zależy od kontekstu). 


😱 Może się zdziwisz, ale podobnie jest z Twoim rozwojem. Co ma większy sens przeczytanie kilku książek szczegółowo opisujących jakieś procesy, o których nie masz pojęcia na starcie, bo nigdy ich nie doświadczyłeś czy doświadczenie czegoś nawet, jeśli nie rozumie się wszystkiego, ale osiąga się już pewien wynik i zaczyna coraz lepiej rozumieć, jakie szczegóły warto w ogóle zgłębiać? Zastanów się ;) 👀


Na koniec ostatnia bardzo ważna rzecz i zaczynamy. Mam prośbę 🙏 Nie porównuj się z innymi. Podczas DWthon każdy jest wygrany. Sam fakt, że tutaj jesteś już jest pewnym krokiem, który został przez Ciebie postawiony, teraz czas na kolejne. Pomyśl o tym, co umiesz w trakcie startu DWthon i każdego dnia zwracaj uwagę na nową rzecz, o której udało Ci się dowiedzieć, czy nową umiejętność, którą udało Ci się zdobyć. Mierz swoją deltę rozwoju, a nie innych :) 

Zaczynamy! 

Najpierw obejrzyj wideo poniżej, aby je zobaczyć postaw kursor w komórce poniżej i naciśnij `SHIFT` + `ENTER`



### Uwaga! 

Jeśli napotkasz trudności podczas wykonywania zadań z tego notebooka, to koniecznie napisz w kanale [dwthon_day1](https://bit.ly/30sh4El)

Pamiętaj, aby szczegółowo doprecyzować, z czym masz problem. Najlepiej wrzuć screen z kodem swoim lub błędem, który widzisz i napisz, którego zadania dotyczy :) 

In [1]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/oZsGy41Jgs4" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

### Krok po kroku 

Nagrałem dla Ciebie także materiał wideo "krok po kroku". W wideo poniżej znajduje się dokładnie to, co w tym notebooku tylko tłumaczę wszystko, aby ułatwić Ci pracę i zrozumienie zagadnień i zadań :) Obejrzyj, jeśli potrzebujesz takiego dodatkowego wsparcia. 


In [2]:
%%html
<iframe style="height:500px;width:100%" src="https://www.youtube.com/embed/NMAYudBMe8Q" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>

## Biblioteki

In [4]:
import pandas as pd

## Wczytujemy dane

Wystarczy jedna linijka kodu, aby wczytać prawie milion wierszy w sekundę 👍

In [5]:
df = pd.read_hdf("../input/data.h5")

df.shape

(820906, 9)

Zbadajmy krok po kroku co tu się stało:
- `pd` to jest skrót, aby odwołać się do biblioteki `pandas` (można powiedzieć "ksywka")
- `pd.read_hdf` wczytuje nasz zbiór danych w formacie hdf/h5 (to jest specjalny format danych, którego dość często używamy, pomińmy teraz szczegóły i wyjaśnienia, dlaczego właśnie taki forma)
- wyniki przypisujemy do zmiennej `df`

#### Czym jest wynik?

Wynikiem jest `DataFrame`. Tak w `pandas` nazywa się tablica. Jak każda dobra tablica, `DataFrame` ma wiersze oraz kolumny.


`df.shape` - zwraca nam "kształt", czyli ile jest wierszy i kolumn w naszej (wczytanej) tablicy. W tym przypadku mamy `(820906, 9)`, czyli **820 906** wierszy oraz **9** kolumn.

## Szybki przegląd

Dwie najważniejsze funkcje na początek to:
- `pd.head()` - zwraca nam pierwsze wiersze. Domyślnie zwraca 5 pierwszych wierszy.
- `pd.sample()` - zwraca losowe wiersze, domyślnie jeden (`pd.sample(10)` zwróci 10 losowych wierszy)

Obie funkcje są przydatne, chociaż warto przyznać, że patrzenie na losowe wiersze (uruchamiając komórkę np. 5-10 razy) może wnieść znacznie więcej wartości dodanej.

Spróbuj uruchomić komórkę 10 razy. Spędź 1-2 minutę na to, aby zapoznać się, jakie wartości występują w poszczególnych kolumnach.



### 💡 Podpowiedź 
Zwykle uruchamiamy komórkę `SHIFT` + `ENTER`. Wynikiem jest uruchomienie komórki i "przeskoczenie" do kolejnych. Czasem chcemy po prostu uruchomić i zostać w tym samym miejscu. To wtedy możemy użyć `CTRL` + `ENTER`. Spróbuj :) 

In [6]:
df.sample(10)

order_id  customer_id  product_id  quantity  price_unit  price_total  \
33548       2421         1074          64         2         254          508   
780505     50771          921         689        12         125         1500   
417066     26023         1433          16         2         595         1190   
516777     33398         1284        1856         2        3995         7990   
448684     28418          577         262        21         125         2625   
418227     26101          391          66        24          55         1320   
421614     26377          311        1671         6         295         1770   
168791     12006         2719        2744         4         415         1660   
554366     36080         5162        1571        60          21         1260   
584524     38363         4506          97         9         425         3825   

        country_id          order_date  is_canceled  
33548            0 2011-01-11 12:55:00        False  
780505           0 2010-11-18 14:07:00        False  
417066           0 2009-12-07 13:00:00        False  
516777           0 2010-04-01 14:52:00        False  
448684           0 2010-01-20 14:04:00        False  
418227           0 2009-12-08 09:35:00        False  
421614           0 2009-12-10 09:44:00        False  
168791           0 2011-06-22 10:28:00        False  
554366           0 2010-05-12 12:32:00        False  
584524           0 2010-06-10 16:25:00         True

😯 Może zdziwić Cię ta kolejność, że najpierw patrzymy na dane i dopiero w drugim kroku próbujemy zinterpretować nazwy kolumn.

Różnie to w życiu bywa, ale z mojego doświadczenia wynika, że popatrzeć 👀 chwilę dłużej na dane opłaca się bardzo! Nawet kilka minut rozważania i patrzenia na dane czasem może zaoszczędzić dni, tygodnie czy nawet miesiące. Często nie zdajemy sobie sprawy, jak bardzo ludzie, którzy pracują na co dzień z pewnymi danymi są zabiegani i nie mają czasu sprawdzić, jak to jest naprawdę, co mówią fakty. Wynikiem tego jest "rozjazd" pomiędzy teorią a stanem faktycznym.

## Sklep


Jak działa każdy sklep?

Są klienci (`customer_id`), którzy kupują produkty (`product_id`) za cenę (`price_unit`) w ilości (`quantity`) i to wszystko łączy się w zamówienie (`order_id`). Dodatkowo mamy czas, kiedy zakup się odbył `order_date`.

#### Przykład.
Liczba/nazwy są losowe. Używam tego przykładu, aby pobudzić Twoją wyobraźnię.


Zośka (`customer_id=1`) składa zamówienie (`order_id`) 8 marca o 10:00:00 (`order_date=08-03-2021 10:00:00`), które składa się z następujących produktów:
- **chleb** (`product_id=1`) jedna sztuka (`quantity=1`) za 3 zł (`price_unit=300`)
- **pomidory** (`product_id=2`) 5 sztuk (`quantity=5`) za 2,55 zł/szt (`price_unit=255`) razem **2,55 * 5 = 12,75 zł** (`price_total=1275`)
  
  
Jako konsument masz prawo do zwrotu. Jeśli zwracasz (w tych danych zwraca się całe zamówienie, nie pojedynczy produkt, tak po prostu jest, nie myśl dlaczego). W takiej sytuacji będzie: `is_canceled=True`. W tym przypadku mówimy, że zamówienie zostało anulowane.


### ☝️ Zwróć uwagę, że `price` trzymamy w groszach.  
*Z bardzo prostego powodu, aby nie mieć kłopotów z zaokrąglaniem. Schodzimy na "fizyczny poziom" naszego świata. Najmniejszą jednostką pieniądza jest "grosz" (cent itd). Dlatego użyjmy tego myślenia też tutaj i zaoszczędzimy sobie sporo nerwów (trochę więcej kontekstu możesz znaleźć np. [tutaj](https://bit.ly/2PI9oMt). Nasze podejście jest 80/20 (80% wartości staramy się wygenerować przy 20% wysiłku), pamiętasz? Dlatego użyjmy bardzo prostego mechanizm, bez komplikowania sobie życia.*


## Od czego zacząć analizę?

Jeśli pracujesz z projektem biznesowym to, co chcesz znaleźć jak najszybciej? Odpowiedź niemal zawsze jest ta sama niezależnie od kultury i czasu - "pieniądze". To jest bardzo prosta reguła, przy pomocy której można dostarczać dużo wartości dodanej firmom i zarabiać jako Data Scientist nawet **23 000 zł** lub więcej miesięcznie w Polsce (o tym więcej mówiłem w tym [filmiku](https://www.youtube.com/watch?v=hxewhurwW3E&t)).

Umiejętność dostarczania wartości "biznesowi" jest fajna też z innego względu. Dzięki temu pracodawca szuka Ciebie, a nie Ty biegasz za pracodawcą. Możesz myśleć, że żartuję, ale jednak mówię prawdę 😉. Zorganizuj swoje życie w taki sposób, żeby to firmy Ciebie potrzebowały (a nie odwrotnie). Postaram się Ci w tym pomóc, a przynajmniej przekażę, jak to działa u mnie 🤝. To zacznijmy już analizę ;)

## Przychody
Policzmy na początek, ile pieniędzy zarobił sklep. Dlaczego to jest ważne? Bo jeśli sklep generuje np. 100 zł, to nawet jak wymyślimy bardzo genialny system usprawnienia, to co z tego?  

W inną stronę, jeśli sklep generuje 10 milionów lub więcej, to wtedy naprawa o jeden procent to już jest dużo (100 tys, prawda?). Jeśli takich poprawek zrobisz 10 w ciągu roku, to wtedy Twoja pensja ~300k już nie brzmi jak coś magicznego. Widać przyczynę i skutek.

Trzymajmy się mocno reguły 80/20. Chcemy skupić się na robieniu takich rzeczy, które faktycznie mają znaczący wpływ. Kierując naszą energię życiową we właściwe miejsce, osiągamy znacznie więcej.

In [7]:
df["price_total"].sum()

1816145630

`price_total` to jest kombinacja `price_unit` * `quantity` dla danego produktu.
Innymi słowy, znajdując sumę (funkcja `.sum()`) dla `price_total` dostajemy informację o przychodach tego sklepu. ☝️ Przypomnę, że `price_total` (oraz `price_unit`) trzymamy w "groszach". Właściwie narzućmy od razu kontekst. Sklep znajduje się w Wielkiej Brytanii i tam jest funt szterling £. Jeden funt dzieli się na 100 pensów. Innymi słowy `price` trzymamy w `pensach`.

#### Dobra, to ile tam wyszło?

Jeśli przyjrzysz się, to zobaczysz, że... **18 161 456.30**, czyli 18+ mln funtów 🧐. Brzmi że mamy, co optymalizować, to bardzo dobrze!


Sprawdźmy teraz, ile pieniędzy nam uciekło (zwroty). W tym przypadku mamy zostawić tylko takie wiersze, które mają `is_canceled = True` i następnie obliczyć sumę.

In [ ]:
df[ True == df.is_canceled  ]["price_total"].sum()

In [10]:
df[ True == df.is_canceled  ]['price_total']

152        465
233       1980
234        672
235        672
236        672
          ... 
820531     165
820532     420
820533     295
820534    1275
820727     195
Name: price_total, Length: 17943, dtype: int32

Wychodzi na to, że **719 739.73** funtów trzeba było zwrócić. Duuuuuuużo!  
To co już warto zapamiętać to to, że z tym można popracować i np. lepiej zrozumieć, które produkty są zwracane lub którzy klienci zwykle zwracają lub jeszcze inne rzeczy!


Zbadajmy jeszcze kilka spraw. Ważne parametry:
- ile mamy klientów?
- ile mamy produktów?


W pandas mamy funkcję `nunique()` (☝️ zwróć uwagę na początku na `n`), to zwraca unikalną liczbę wartości. Zobaczmy to w działaniu.

In [11]:
df["customer_id"].nunique()

5879

Zwróciło nam 5879, to oznacza, że w kolumnie `customer_id` jest tyle unikalnych wartości. Każda wartość jest przypisywana do konkretnej osoby. Dlatego unikalna liczba wartości odpowiada na pytanie, ile mam klientów.

## Zadanie 1.1

Już wiesz, jak znaleźć liczbę klientów, teraz trzeba zrobić to samo dla produktów. Chcemy odpowiedzieć na pytanie, ile produktów mamy w sklepie?


### 💡 Podpowiedzi
1. Zamiast zawartości `## YOUR CODE HERE` wpisz kod
2. Pewnie chcesz użyć `product_id`.
3. Zainspiruj się komórką wyżej, gdzie znaleźliśmy liczbę unikalnych `customer_id`.

In [14]:
df['product_id'].nunique()

3878

## Zadanie 1.2
Skoro już wiemy, ile mamy produktów, to sprawdźmy jeszcze, ile ma zamówień.

💡 Podpowiedzi
1. Zamiast zawartości ## YOUR CODE HERE wpisz kod
2. Pewnie chcesz użyć `order_id`.

In [19]:
df.order_id.nunique()

43908

To już tyle jako plan minimum. Jak wykonasz zadanie, należy zrobić:
1. Zrzut ekranu na którym widać rozwiązania, wtedy dostaniesz bonus  :)  
2. Wrzucić rozwiązanie na slacku do pokoju [dwthon_day1_done](https://bit.ly/3brCUyp)
3. Dostać bonus za dobrą robotę 💪 i jutro będzie kolejna porcja.

Uwaga! Jeśli masz problem z jakimś zadaniem, czegoś nie wiesz, to pamiętaj, że możesz zadawać pytania! Na tym polega nauka :) Pytania związane z 1 dniem DWthon zadawaj w kanale [dwthon_day1](https://bit.ly/30sh4El).

## Dodatkowe zadania

Nie musisz ich wykonywać, ale jak czujesz "głód" i chcesz więcej, to wtedy działaj śmiało 👍.


### 🧠 Włączmy myślenie krytycznie. 
Ta umiejętność jest bardzo potrzebna przy analizowniu danych (i ogólnie przydaje się w życiu).

Powyżej było zadanie, ile jest produktów w sklepie, prawda? Ale czy jesteśmy w stanie odpowiedzieć na to pytanie? Pomyśl...


Tak naprawdę, nie! Bo nie wiemy, ile jest produktów w sklepie. To co wiemy, jakie produkty sprzedały się co najmniej raz! To jest ogromna różnica, bo może być jeszcze tysiąc innych produktów (których nie ma w naszych danych/zamówieniach), które generują koszty, ale nie sprzedają się w ogóle. Dlatego bardzo uważnie rozważaj, jakie masz dane ;). 


Policzmy teraz "średni rachunek". Wiemy, jakie są przychody. Też wiemy, ile było zamówień, prawda? To podzielmy jedno przez drugie.

In [ ]:
df["price_total"].sum() / df["order_id"].nunique()

W tak łatwy sposób dowiedzieliśmy się, że średni rachunek wynosi **413.63** funta. Dość dużo, prawda?

Podpowiem Ci, że w tym przypadku chodzi o sklep z prezentami. Jak kupujesz prezenty dla swojej rodziny czy znajomych  (np. na Święta), to jest szansa, że wydaje się podobne pieniądze ;)

## 😇 Włączmy myślenie kreatywne

Jakie jeszcze 5-10 metryk warto byłoby sprawdzić? Pomyśl na spokojnie, na pewno wymyślisz, daj sobie chwilę czasu. Po prostu masz zrobić dla siebie przestrzeń, porozważać, być może pójść na krótki spacerek, może szybki zimny prysznic 🚿, medytacja lub coś innego, co pobudza Cię do kreatywnego myślenia. 

## 🤝🗣️ Pomyśl o kooperacji i komunikacji

Podziel się pomysłami, co już udało się Ci wymyślić w poprzednim kroku i podziel się tym na slacku [dwthon_day1_ideas](https://bit.ly/3c9ocLB). Zaangażuj się w dyskusję i zobacz, jak w prosty sposób wytwarza się dużo pomysłów przy współpracy z innymi.

## Twój Feedback i delta rozwoju 

Daj mi znać koniecznie, co sprawiło Ci przyjemność, co trudność oraz czego udało się nauczyć w tej krótkiej [ankiecie](https://bit.ly/3pQVBBP). Pozwoli mi ona ulepszyć zadania, a Tobie lepiej trakować swój rozwój: win-win 😇 

## Widzimy się jutro!  🙌

## 🕵️‍♀️ 3 fakty o DataWorkshop dla głodnych wiedzy 😎


#### 1. 💪 Jesteśmy praktykami 

Uczymy Cię na podstawie naszego codziennego doświadczenia <3, a nie teorii wyczytanej w książkach. 
Podstawą naszych działań jest Data Science & Machine Learning - pomagamy firmom na różnym poziomie wdrażać skuteczne modele Machine Learning. A więc uczysz się tego, co działa w praktyce i od praktyków :) Jeśli zależy Ci na zdobywaniu konkretnych umiejętności cenionych na ryku, to jesteś we właściwym miejscu. 

> [Tutaj](https://bit.ly/3wBzPG2)  możesz posłuchać lub poczytać o jednym z naszych projektów. 

Masz pytania na temat współpracy? 

👉 Pisz śmiało na adres: ula@dataworkshop.eu lub hello@dataworkshop.eu 


#### 2. 🤝 Pomagamy zostać specjalistą / specjalistką  

Jeśli chcesz związać się zawodowo z Data Science i/lub Machine Learning, to koniecznie poczytaj o naszych kompleksowym programie nauki, który pomógł już niejednej osobie zmienić pracę, być lepszym w swojej pracy, wzbogacić kompetencje niezależnie od zajmowanego stanowiska. 

> [**Data Science & Machine Learning w praktyce**](
https://bit.ly/3PWyiB6) -  to najlepszy kurs, który pomoże Ci postawić pierwsze kroki lub pomóc Ci wskoczyć na kolejny poziom umiejętności branżowych. 100% praktyki, uczysz się tego, co najważniejsze i czego nie znajdziesz w książkach. 

Masz pytania na temat kursu? 

👉 Pisz śmiało na adres: ela@dataworkshop.eu lub hello@dataworkshop.eu 


#### 3. 🤝 Pomagamy stawiać pierwsze kroki 


Wiemy, że czasem Python wydaje się być przeszkodą w nauce ML... Cóż dobra wiadomość jest taka, że nie jest to przeszkoda nie do pokonania i w tym też możemy Ci pomóc. 

Kursów i materiałów z Python jest mnóstwo, ale czy warto uczyć się wszystkiego? 

Może nie wystarczyć czasu, a możesz nauczyć się tego, co nie jest aż tak istotne jeśli chodzi o tematykę Data Science & ML. 

Wypełniliśmy tę lukę i Vladimir stworzył kurs 

> [**Analiza danych w Python**](https://bit.ly/3Tl6M3l) - to materiał, dzięki któremu uczysz się tego w Python, co jest kluczowe w kontekście analizy danych i "data science". 

Masz pytania na temat tego kursu? 

👉 Pisz śmiało na adres: ela@dataworkshop.eu lub hello@dataworkshop.eu 
